src/validators/keystores/hashi_vault.py

```diff
@@ -38,52 +44,57 @@ def from_settings(cls) -> 'HashiVaultConfiguration':
+            f'/v1/{self.engine_name}/{location}/{key_path}',
         )
 
+    def prefix_url(self, keys_prefix: str) -> str:
+        """An URL for Vault secrets engine location that holds prefixes for keys."""
+        keys_prefix = keys_prefix.strip('/')
+        # URL is used for listing, so it lists metadata
+        return self.secret_url(keys_prefix, location='metadata')
+
+
+@dataclass
+class HashiVaultKeysLoader(metaclass=abc.ABCMeta):
+    config: HashiVaultConfiguration
+    input_iter: Iterator[str]
+
+    def session(self) -> AsyncContextManager:
+        return ClientSession(
+            timeout=ClientTimeout(HASHI_VAULT_TIMEOUT),
+            headers={'X-Vault-Token': self.config.token},
+        )
 
-class HashiVaultKeystore(LocalKeystore):
     @staticmethod
-    async def load() -> 'HashiVaultKeystore':
-        """Extracts private keys from the keystores."""
-        hashi_vault_config = HashiVaultConfiguration.from_settings()
+    def merge_keys_responses(keys_responses: list[Keys], merged_keys: Keys) -> None:
+        for keys in keys_responses:
+            for pk, sk in keys.items():
+                if pk in merged_keys:
+                    logger.error('Duplicate validator key %s found in hashi vault', pk)
+                    raise RuntimeError('Found duplicate key in path')
+                merged_keys[pk] = sk
 
-        parallelism = hashi_vault_config.parallelism
-        key_paths = hashi_vault_config.key_paths
-        merged_keys = Keys({})
+    async def load_into_merged(self, merged_keys: Keys) -> None:
+        while key_chunk := list(itertools.islice(self.input_iter, self.config.parallelism)):
+            await self.process_keys_chunk(key_chunk, merged_keys)
 
-        key_paths_iter = iter(key_paths)
-        while key_chunk := list(itertools.islice(key_paths_iter, parallelism)):
-            async with ClientSession(
-                timeout=ClientTimeout(HASHI_VAULT_TIMEOUT),
-                headers={'X-Vault-Token': hashi_vault_config.token},
-            ) as session:
-                keys_responses = await asyncio.gather(
-                    *[
-                        HashiVaultKeystore._load_hashi_vault_keys(
-                            session=session,
-                            secret_url=hashi_vault_config.secret_url(key_path),
-                        )
-                        for key_path in key_chunk
-                    ]
-                )
-                for keys in keys_responses:
-                    for pk, sk in keys.items():
-                        if pk in merged_keys:
-                            logger.error('Duplicate validator key %s found in hashi vault', pk)
-                            raise RuntimeError('Found duplicate key in path')
-                        merged_keys[pk] = sk
-        return HashiVaultKeystore(merged_keys)
+    @abc.abstractmethod
+    async def process_keys_chunk(self, input_chunk: list[str], merged_keys: Keys) -> None:
```

Comment:
It's unclear what is `input_chunk`. It has different semantics for different loaders. I suggest to add `load` abstract method which has no args. So loader would work like this:
```
loader = HashiVaultBundledKeysLoader(hashi_config)
keys = loader.load()
```

Reply:
Refactored in https://github.com/stakewise/v3-operator/pull/439/commits/7def6db7a02737dc41f3c33482fcf03b12d76521, now each chunk variable is named sufficiently: `key_chunk` and `prefix_chunk`